Added mailhog for e2e tests

9larsons opened the pull request:
ref https://linear.app/ghost/issue/PROD-2515/
- added Mailhog to the e2e package to support email related functionality; this is necessary piece of member signups
- added a member sign up flow test

This set of changes is to enable the ability to do full e2e member signup and attribution tests. In order to do that, we need email support. Here we add Mailhog, which is used in our development Docker setup, in order to accomplish this.

We are fairly tightly coupled to Mailhog here as we explore this option and how it works for testing before moving to an adapter pattern like we use elsewhere.

9larsons committed:
Added mailhog for e2e tests

cmraible reviewed (commented):
Nice, surprisingly clean and succinct for a first draft! I don't think we're far off here, the biggest open question to me is how we do the mailhog stuff in such a way that it plays nicely with parallelism.

9larsons committed:
Merge branch 'main' into e2e-add-email-signup

9larsons committed:
Merge branch 'main' into e2e-add-email-signup

9larsons committed:
Fixed linting

9larsons commented on e2e/tests/public/member-signup-with-email.test.ts:
```diff
@@ -0,0 +1,102 @@
+    test.beforeEach(async () => {
+        mailhog = new MailhogClient();
+        // Clear any existing emails before each test
+        await mailhog.deleteAllMessages();
```
Good reminder. Since we do lookup via email, I was planning to not remove any messages and we can use generators/helpers to make sure we're using unique email addresses.

9larsons committed:
Removed partial tests

9larsons committed:
Fixed tests, added helpers

9larsons committed:
Created sign up flow

9larsons committed:
Wrapped faker generators into a helper

9larsons committed:
Updated tests to be more readable

9larsons committed:
Refactored use of faker to use generators

ARguably we should get rid of utils for generators, but these are specific to the factories

9larsons commented on e2e/tests/public/member-signup.test.ts:
```diff
@@ -0,0 +1,87 @@
+        const isClosed = await successPage.isPortalClosed();
+        expect(isClosed).toBe(true);
+    });
+});
```
Nah I'll toss them.

9larsons committed:
Updated ports

9larsons committed:
Merge branch 'main' into e2e-add-email-signup

9larsons committed:
Fixed faker call

ibalosh commented on e2e/helpers/email/MailhogClient.ts:
```diff
@@ -0,0 +1,192 @@
+import baseDebug from '@tryghost/debug';
```
I agree, I would do though some small changes towards making this easier to maintain later:

Add interface for the `EmailClient` so that in tests, we don't need to mention mailhog client that much, but rather emailClient.

```typescript
let emailClient: EmailClient;

test.beforeEach(async () => {
       emailClient = new MailhogClient();
});
```

Also, we should separate out responsibilities that are not of the email client from it. Client should retrieve, search, delete email.

Email client should be in charge of:

* getMessages
* searchForMessages(filterBy: recipient | emailSubject, options: numberOfEmails, searchTimeLimit -> waitFor
*  deleteMessages(options)

Email utils 

* getMessageTextContent(message: EmailMessage)
* getMessageHtmlContent(message: EmailMessage)
* extractLink(message, magicLinkRegexp)

Parts which should not be part of email client:

* waitForMagicLink - since it's a mix of wait and extract, and too specific for email client
* completeSignupVerification - mix of page and wait

It's ok, to show some of the workflow in test:

```
const {emailAddress} = await signupViaPortal(page);
const message = await this.emailClient.searchForMessage(emailAddress, timeoutMs);
const magicLink = this.extracLinkFromMessage(message, magicLinkRegexp);

const signupVerificationPage = new SignupVerficitationPage(page);
await page.goto(magicLink);
expect(signupVerificationPage.loaded()).toBeTrue();
```

ibalosh commented on e2e/tests/public/member-signup-with-email.test.ts:
```diff
@@ -0,0 +1,35 @@
+        expect(emailBody).toContain('complete the signup process');
+
+        const magicLink = mailhog.extractMagicLink(message);
+        expect(magicLink).toBeTruthy();
```
we can skip all three checks, by just visiting the extracted magic link? If it works, we verified it. Current verification goes into implementation details with checking the content of the magic link.

9larsons replied:
That's fair. I was on the fence w/ this one. I suppose it makes sense to simply visit and make sure it works, though that's covered by a Portal test. All I really twanted to confirm here was that the email was received (looking appropriate helps). Where does that leave us? Do we feel this is unnecessary and should simply have a signup flow test instead?

ibalosh reviewed (changes requested):
Hey @9larsons 

I took a look at this, and I like the approach, I would pretty much take the exact same route as you did.

One thing that I would improve though is the mailhog client, it does bit too much now, it's mixing things.
 
If you have too much stuff on your plate though, let me know, and I can help out on cleanup there, and we can roll this out earlier.

9larsons committed:
Updated accountButton

9larsons committed:
Rm unnecessary comments

9larsons committed:
Rm generators file abstraction

ibalosh reviewed (approved):
let's add it in, and we can cleanup later client. I would only move:

```
const magicLink = await this.waitForMagicLink(email, timeoutMs);

await page.goto(magicLink);
await page.waitForLoadState('networkidle');
```

out of mailhog client to the test, not to clutter client with playwright

9larsons committed:
Updated test file name, rm playwright from mailhogclient

9larsons committed:
Updted magic link handling
